### PR TITLE
Adding test cases

### PR DESCRIPTION
**Pull request by ggaurav10**
**What this PR does / why we need it**:
Adding test cases

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator

```

### CODE REVIEW

File: controllers/hpa_util_test.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+				return fmt.Errorf("Error: Number of HPAs expected: 1; found %v", num)
+			}, timeout).Should(Succeed())
+
+			// Create new HPA for same HVPA
+			newHpa, err := getHpaFromHvpa(instance)
+			Expect(err).NotTo(HaveOccurred())
+			err = c.Create(context.TODO(), newHpa)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Eventually one of the HPAs should be garbage collected
```
Comment (amshuman-kr) [inline controllers/hpa_util_test.go]: Please add an assertion here to verify that there are indeed two HPAs now.

File: controllers/hpa_util_test.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+			Eventually(func() error {
```
Comment (amshuman-kr) [inline controllers/hpa_util_test.go]: Reuse the func from [above](https://github.com/gardener/hvpa-controller/pull/41/files#diff-fc45a10d64187f8097a44a04a80d30d9R80)?

File: controllers/vpa_util_test.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+				return fmt.Errorf("Error: Number of VPAs expected: 1; found %v", num)
+			}, timeout).Should(Succeed())
+
+			// Create new VPA for same HVPA
+			newVpa, err := getVpaFromHvpa(instance)
+			Expect(err).NotTo(HaveOccurred())
+			err = c.Create(context.TODO(), newVpa)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Eventually one of the VPAs should be garbage collected
```
Comment (amshuman-kr) [inline controllers/vpa_util_test.go]: Please add an assertion here to verify that there are indeed two VPAs now.

File: controllers/vpa_util_test.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+			Eventually(func() error {
+				num := 0
```
Comment (amshuman-kr) [inline controllers/vpa_util_test.go]: Reuse the func [above](https://github.com/gardener/hvpa-controller/pull/41/files#diff-550ccb08665a59804d8d7d631e7b2c29R80)?

File: controllers/hpa_util_test.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+					}
+				}
+				if num == 1 {
+					return nil
+				}
+				return fmt.Errorf("Error: Number of HPAs expected: 1; found %v", num)
+			}, timeout).Should(Succeed())
+
+			// Create new HPA for same HVPA
+			newHpa, err := getHpaFromHvpa(instance)
```
Comment (amshuman-kr) [inline controllers/hpa_util_test.go]: Can you please add a couple more variations where the newHPA is tweaked to not have owner reference and/or `GenerateName` filled but of course still with the same labels?

Also, one more negative case, to create an HPA without labels and make sure it does *not* get adopted?

The same comments apply for VPA too.

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
Requested some small changes and a 2-3 variations if the tests to cover slightly different scenarios (including a negative one).

File: controllers/hpa_util_test.go (hunk 1 above)
Comment (ggaurav10) [reply controllers/hpa_util_test.go]: There could be a race condition in that case; the HPA could be garbage collected before any assertion placed here. I think the assertion that `Create()` of `newHpa` is successful is enough. WDYT?

File: controllers/vpa_util_test.go (hunk 3 above)
Comment (ggaurav10) [reply controllers/vpa_util_test.go]: There could be a race condition in that case; the VPA could be garbage collected before any assertion placed here. I think the assertion that Create() of newVpa is successful is enough. WDYT?

File: controllers/hpa_util_test.go (hunk 1 above)
Comment (amshuman-kr) [reply controllers/hpa_util_test.go]: Understood. Troubles of not having fake client. Alright. That's ok, then.

File: controllers/hpa_util_test.go (hunk 5 above)
Comment (ggaurav10) [reply controllers/hpa_util_test.go]: >Can you please add a couple more variations where the newHPA is tweaked to not have owner reference

`getHpaFromHvpa()` returns `newHpa` without `owerRef` already.
Will add rest of the variations.

**Review by amshuman-kr — APPROVED** [review]
Thanks for the modularisation and variations in the test cases! LGTM

**Pull request merged**